pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -572,16 +576,69 @@ func matchFailureDomains(gMS *machineapi.MachineSet, rMS machineapi.MachineSet,
 	rMS_providerconfig, err := cpms.NewProviderConfigFromMachineSpec(logr, rSpec, infrastructure)
 	if err != nil {
 		logger.WithError(err).Errorf("unable to parse remote MachineSet %v provider config", rMS.Name)
-		return false, err
+		return false, nil, err
 	}
 	gMS_providerconfig, err := cpms.NewProviderConfigFromMachineSpec(logr, gSpec, infrastructure)
 	if err != nil {
 		logger.WithError(err).Errorf("unable to parse generated MachineSet %v provider config", gMS.Name)
-		return false, err
+		return false, nil, err
 	}
 	fd1 := rMS_providerconfig.ExtractFailureDomain()
 	fd2 := gMS_providerconfig.ExtractFailureDomain()
-	return fd1.Equal(fd2), nil
+
+	if rMS_providerconfig.Type() == "AWS" {
+		subnets := []*ec2.Subnet{}
+		awsactuator := actuator.(*AWSActuator)
+
+		if aws.StringValue(fd1.AWS().Subnet.ID) == "" {
+			subnets, err = getSubnetIDByFilter(fd1.AWS().Subnet, *awsactuator)
+			if err != nil {
+				return false, nil, err
+			}
+		}
+		subnets, err = awsactuator.filterPublicSubnets(subnets, pool)
+		if err != nil {
+			return false, nil, err
+		}
+
+		if aws.StringValue(subnets[0].AvailabilityZone) != fd2.AWS().Placement.AvailabilityZone {
+			return false, subnets, nil
+		} else {
+			return true, subnets, nil
+		}
+	}
+
+	return fd1.Equal(fd2), nil, nil
+}
+
+func getSubnetIDByFilter(subnet *v1.AWSResourceReference, actuator AWSActuator) ([]*ec2.Subnet, error) {
+	for _, filter := range *subnet.Filters {
+		if filter.Name == "tag:Name" { // FIXME use all filters or just this one?
```

Comment:
All filters. The point is to find the same subnet(s) that AWS would when fed this filter from the Machine itself. So just `req := &ec2.DescribeSubnetsInput{Filters: subnet.Filters}` and go. No outer loop.

Reply:
> No outer loop.

Okay, so those filters are different types internally (scalar vs pointer types in the leaves of the structs) so you will still have to have the outer loop to convert.